tests-unified/GeniusYield/Test/Unified/BetRef/PlaceBet.hs

```diff
@@ -0,0 +1,257 @@
+    , mkPrivnetTestFor_ "Balance checks after placing first bet - privnet" firstBetTest
+    , mkTestFor "Balance checks with multiple bets" multipleBetsTest
+    , mkPrivnetTestFor_ "Balance checks with multiple bets - privnet" multipleBetsTest
+    , mkTestFor "Not adding atleast bet step amount should fail" $ mustFail . failingMultipleBetsTest
+    , mkPrivnetTestFor' "Not adding atleast bet step amount should fail - privnet" GYDebug setup $
+        handleError
+          (\case
+              GYBuildTxException GYBuildTxBodyErrorAutoBalance {} -> pure ()
+              e -> throwError e
+          )
+        . failingMultipleBetsTest
+    ]
+  where
+    mkPrivnetTestFor_ = flip mkPrivnetTestFor setup
+    firstBetTest :: GYTxGameMonad m => TestInfo -> m ()
+    firstBetTest = firstBetTrace (OracleAnswerDatum 3) (valueFromLovelace 20_000_000) . testWallets
+    multipleBetsTest :: GYTxGameMonad m => TestInfo -> m ()
+    multipleBetsTest TestInfo{..} = multipleBetsTraceWrapper 400 1_000 (valueFromLovelace 10_000_000)
+      [ (w1, OracleAnswerDatum 1, valueFromLovelace 10_000_000)
+      , (w2, OracleAnswerDatum 2, valueFromLovelace 20_000_000)
+      , (w3, OracleAnswerDatum 3, valueFromLovelace 30_000_000)
+      , (w2, OracleAnswerDatum 4, valueFromLovelace 50_000_000)
+      , (w4, OracleAnswerDatum 5, valueFromLovelace 65_000_000 <> valueSingleton testGoldAsset 1_000)
+      ]
+      testWallets
+    failingMultipleBetsTest :: GYTxGameMonad m => TestInfo -> m ()
+    failingMultipleBetsTest TestInfo{..} = multipleBetsTraceWrapper 400 1_000 (valueFromLovelace 10_000_000)
+      [ (w1, OracleAnswerDatum 1, valueFromLovelace 10_000_000)
+      , (w2, OracleAnswerDatum 2, valueFromLovelace 20_000_000)
+      , (w3, OracleAnswerDatum 3, valueFromLovelace 30_000_000)
+      , (w2, OracleAnswerDatum 4, valueFromLovelace 50_000_000)
+      , (w4, OracleAnswerDatum 5, valueFromLovelace 55_000_000 <> valueSingleton testGoldAsset 1_000)
+      ]
+      testWallets
+
+-- -----------------------------------------------------------------------------
+-- Super-trivial example
+-- -----------------------------------------------------------------------------
+
+-- | Trace for a super-simple spending transaction.
+simplSpendingTxTrace :: GYTxGameMonad m => Wallets -> m ()
+simplSpendingTxTrace Wallets{w1} = do
+  gyLogDebug' "" "Hey there!"
+  -- balance assetion check
+  withWalletBalancesCheckSimple [w1 := valueFromLovelace (-100_000_000)] . asUser w1 $ do -- TODO: w1 is the wallets that gets all funds for now
+    skeleton <- mkTrivialTx
+    gyLogDebug' "" $ printf "tx skeleton: %s" (show skeleton)
+
+    -- test itself
+    txId <- buildTxBody skeleton >>= signAndSubmitConfirmed
+    gyLogDebug' "" $ printf "tx submitted, txId: %s" txId
+
+-- Pretend off-chain code written in 'GYTxMonad m'
+mkTrivialTx :: GYTxMonad m => m (GYTxSkeleton 'PlutusV2)
+mkTrivialTx = do
+  addr <- fmap (!! 0) ownAddresses -- FIXME:
```

Comment:
The fact that `ownChangeAddress` is not `Maybe GYAddress` but `GYAddress`, should we modify `ownAddresses` to instead be `NonEmpty GYAddress` rather than `[GYAddress]`?